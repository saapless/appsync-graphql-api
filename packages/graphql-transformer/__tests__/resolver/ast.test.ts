import {
  _call,
  _chain,
  _export,
  _func,
  _id,
  _obj,
  _prop,
  _return,
  MemberExpression,
  NodeKind,
  printAST,
} from "../../src/resolver";

describe("template ast parser", () => {
  describe("utility functions", () => {
    it("_chain => build MemberExpression", () => {
      const ast = _chain("ctx.args.id");

      expect(ast).toEqual({
        _kind: NodeKind.MEMBER_EXPRESSION,
        object: {
          _kind: NodeKind.MEMBER_EXPRESSION,
          object: {
            _kind: NodeKind.IDENTIFIER,
            name: "ctx",
          },
          property: {
            _kind: NodeKind.IDENTIFIER,
            name: "args",
          },
          computed: false,
          optional: false,
        },
        property: {
          _kind: NodeKind.IDENTIFIER,
          name: "id",
        },
        computed: false,
        optional: false,
      } satisfies MemberExpression);
    });

    it("create function declaration ast", () => {
      const ast = _export(
        _func(
          "request",
          [_id("ctx")],
          _return(_call(_id("get"), [_obj(_prop("key", _obj(_prop("id", _chain("ctx.args.id")))))]))
        )
      );

      expect(ast).toMatchSnapshot();
      expect(printAST(ast)).toMatchSnapshot();
    });
  });
});
