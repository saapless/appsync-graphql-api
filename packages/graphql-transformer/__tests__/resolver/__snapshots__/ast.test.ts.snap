// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template ast parser utility functions create function declaration ast 1`] = `
{
  "_kind": "ExportNamedDeclaration",
  "declaration": {
    "_kind": "FunctionDeclaration",
    "body": {
      "_kind": "BlockStatement",
      "body": [
        {
          "_kind": "ReturnStatement",
          "value": {
            "_kind": "CallExpression",
            "arguments": [
              {
                "_kind": "ObjectExpression",
                "properties": [
                  {
                    "_kind": "Property",
                    "name": "key",
                    "value": {
                      "_kind": "ObjectExpression",
                      "properties": [
                        {
                          "_kind": "Property",
                          "name": "id",
                          "value": {
                            "_kind": "MemberExpression",
                            "computed": false,
                            "object": {
                              "_kind": "MemberExpression",
                              "computed": false,
                              "object": {
                                "_kind": "Identifier",
                                "name": "ctx",
                              },
                              "optional": false,
                              "property": {
                                "_kind": "Identifier",
                                "name": "args",
                              },
                            },
                            "optional": false,
                            "property": {
                              "_kind": "Identifier",
                              "name": "id",
                            },
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            ],
            "callee": {
              "_kind": "Identifier",
              "name": "get",
            },
            "optional": false,
          },
        },
      ],
    },
    "name": {
      "_kind": "Identifier",
      "name": "request",
    },
    "parameters": [
      {
        "_kind": "Identifier",
        "name": "ctx",
      },
    ],
  },
  "specifiers": undefined,
}
`;

exports[`template ast parser utility functions create function declaration ast 2`] = `
"export function request(ctx) {
  return get({ key: { id: ctx.args.id } })
}
"
`;
